### PR TITLE
Implement dockerhub adapter

### PR DESCRIPTION
**Pull request by cd1989**
Signed-off-by: cd1989 <chende@caicloud.io>

Implements registry adapter for DockerHub.

### CODE REVIEW

File: src/replication/ng/adapter/dockerhub/adapter.go [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+}
+
+// Ensure '*adapter' implements interface 'Adapter'.
+var _ adp.Adapter = (*adapter)(nil)
+
+// Info returns information of the registry
+func (a *adapter) Info() (*model.RegistryInfo, error) {
+	return &model.RegistryInfo{
+		Type: model.RegistryTypeDockerHub,
+		SupportedResourceTypes: []model.ResourceType{
```
Comment (ywk253100) [inline src/replication/ng/adapter/dockerhub/adapter.go]: Add "SupportNamespace" field, it should be true

File: src/replication/ng/adapter/dockerhub/adapter.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+				Type:  model.FilterTypeName,
+				Style: model.FilterStyleTypeText,
+			},
+			{
+				Type:  model.FilterTypeTag,
+				Style: model.FilterStyleTypeText,
+			},
+		},
+		SupportedTriggers: []model.TriggerType{
+			model.TriggerTypeManual,
```
Comment (ywk253100) [inline src/replication/ng/adapter/dockerhub/adapter.go]: Should support scheduled trigger either.

File: src/replication/ng/adapter/dockerhub/adapter.go [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+		return false
+	}
+
+	return true
+}
+
+// ConvertResourceMetadata converts the namespace and repository part of the resource metadata
+// to the one that the adapter can handle
+func (a *adapter) ConvertResourceMetadata(meta *model.ResourceMetadata, namespace *model.Namespace) (*model.ResourceMetadata, error) {
+	return meta, nil
```
Comment (kofj) [inline src/replication/ng/adapter/dockerhub/adapter.go]: how to process full repository path like `a/b/c:tag`?

File: src/replication/ng/adapter/dockerhub/adapter.go [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode/100 != 2 {
```
Comment (kofj) [inline src/replication/ng/adapter/dockerhub/adapter.go]: what different  between this and `resp.StatusCode != 200`?

Comment (cd1989) [reply src/replication/ng/adapter/dockerhub/adapter.go]: 200 vs 2xx

File: src/replication/ng/adapter/dockerhub/adapter.go (hunk 3 above)
Comment (cd1989) [reply src/replication/ng/adapter/dockerhub/adapter.go]: For the moment, only Habor -> DockerHub, this suffices, for other cases, let's consider it later after  https://github.com/goharbor/harbor/issues/7361

**Pull request merged**